### PR TITLE
NEW: Implement least-squares + gradient method

### PR DESCRIPTION
**Pull request by carterbox**
## Purpose
Implement the method from "Iterative least-squares solver for generalized maximum-likelihood ptychography" because this method is capable to iterating faster by using certain approximations and avoiding many FFTs.

## Approach
Tried to follow the text of the paper, but deviated when implementing less features than the paper and when the paper didn't make sense to me.

This PR also contains the following features:

- renamed `combined` algorithm to `cgrad`.
- Implemented batch updates for both conjugate gradient and least squares methods
- reconstruct now returns wall time of each iteration
- default field of view is now only padded by a single probe width instead of 1/8 the field of view
- `orthogonalize_eig` is now done on the GPU
- tried to improve update times for probe mode updates in conjugate gradient by caching intensities from other modes 

### Submitter
- [x] Write a helpfully descriptive pull request title.
- [x] Organize changes into logically grouped commits with descriptive commit messages.
- [ ] Document all new functions.
- [ ] Write tests for new functions or explain why they are not needed.
- [ ] Build the documentation successfully
- [x] Use [`yapf`](https://github.com/google/yapf) to format python code.

### Reviewer
- [ ] Actually read all of the code.
- [ ] Run the new code yourself.
- [ ] Write a summary of the changes as you understand them.
- [ ] Thank the submitter.

### CODE REVIEW

**Comment by carterbox** [comment]
@xiaodong-yu asked:
>You said the LS solver avoids FFT, but I don't see any changes in the CUDA kernel; I wonder how and where the FFT is avoided? Are the LS and CG solvers orthogonal in the codebase? For example, if I want to add multi-node support to the CG solver, do I need to also touch the LS solver code?

Yes, the least-squares method is orthogonal to the conjugate-gradient method because it is a different algorithm. They will need to be developed and maintained mostly separately.

File: src/tike/ptycho/solvers/combined.py [hunk 1]
```diff
@@ -64,19 +63,48 @@ def combined(
...
     """Solve the probe recovery problem."""
 
-    # TODO: Cache object patche between mode updates
+    # TODO: Cache object patches between mode updates
+    intensity = [
+        _compute_intensity(op, psi, scan, probe[..., m:m + 1, :, :])[0]
+        for m in range(probe.shape[-3])
+    ]
+    intensity = op.xp.array(intensity)
+
```
Comment (carterbox) [inline src/tike/ptycho/solvers/combined.py:87]: However, this code chunk is a patch to the conjugate-gradient solver which avoids reduces the number of FFTs by caching the detector intensity contribution from the probe modes that are not being updated.

Comment (xiaodong-yu) [reply src/tike/ptycho/solvers/combined.py:87]: Does this mean reducing the number of FFT is a code optimization for both LS and CG solvers, regardless of the LS algorithm?

Comment (carterbox) [reply src/tike/ptycho/solvers/combined.py:87]: The least-squares method intrinsically performs less FFTs per iteration when compared to the conjugate gradient method. This PR also contains a patch which reduces the number of FFTs performed by the conjugate gradient method.

**Comment by carterbox** [comment]
Profiling results from this PR with the scan 145 dataset and a batch size of 128. The first most time is actually spent computing sums. Apparently, we need to focus more on avoiding sums and memory copies.

Profiling results from `cgrad`.
```bash
            Type  Time(%)      Time     Calls       Avg       Min       Max  Name
 GPU activities:   47.51%  537.293s    327466  1.6408ms  3.8400us  5.4749ms  cupy_sum
                    7.01%  79.2596s    125380  632.16us  339.71us  894.30us  void vector_fft<unsigned int=256, unsigned int=16, unsigned int=1, padding_t=6, twiddle_t=0, loadstore_modifier_t=2, layout_t=0, unsigned int, float>(kernel_arguments_t<unsigned int>)
                    7.01%  79.2288s    128380  617.14us  22.015us  1.5684ms  cupy_true_divide__complex64_complex_complex64
                    6.93%  78.3573s    119380  656.37us  377.85us  1.0644ms  void spRadix0256C::kernel1Mem<unsigned int, float, fftDirection_t=-1, unsigned int=16, unsigned int=2, L1, ALL, WRITEBACK>(kernel_parameters_t<fft_mem_radix1_t, unsigned int, float>)
                    6.53%  73.8937s    345391  213.94us  1.7920us  783.26us  cupy_sqrt__float32_float32
                    4.73%  53.4467s    120880  442.15us  1.7600us  1.0594ms  cupy_absolute__complex64_float32
                    4.44%  50.1750s    109678  457.48us  1.8560us  923.29us  cupy_subtract__float32_float32_float32
                    3.24%  36.5880s    116380  314.38us  168.45us  592.38us  cupy_square__float32_float32
                    3.09%  34.9369s    128380  272.14us  10.752us  627.36us  [CUDA memset]
                    3.06%  34.5470s    114131  302.70us  1.1840us  500.38us  cupy_multiply__float32_float32_float32
                    2.64%  29.8983s     93205  320.78us  40.383us  537.28us  [CUDA memcpy DtoD]
                    0.65%  7.31783s    129880  56.343us  1.4070us  277.41us  cupy_multiply__complex64_complex64_complex64
                    0.65%  7.29546s     23175  314.80us  171.58us  652.09us  cupy_add__float_float32_float32
                    0.51%  5.78182s    122380  47.244us  35.680us  251.93us  fwd_patch
                    0.41%  4.68783s     10501  446.42us  1.5680us  1.1281ms  cupy_multiply__complex64_float32_complex64
                    0.38%  4.33252s    117129  36.989us  1.2480us  315.97us  cupy_add__complex64_complex64_complex64
                    0.35%  3.95631s      6000  659.39us  646.04us  899.67us  void spRadix0256C::kernel1Mem<unsigned int, float, fftDirection_t=1, unsigned int=16, unsigned int=2, L1, ALL, WRITEBACK>(kernel_parameters_t<fft_mem_radix1_t, unsigned int, float>)
                    0.24%  2.76490s      6050  457.01us  2.0160us  535.58us  cupy_true_divide__float32_float32_float32
                    0.18%  2.05268s     92455  22.201us  1.4710us  563.93us  cupy_add__float32_float_float32
                    0.17%  1.88739s      6000  314.56us  308.41us  382.59us  cupy_subtract__float_float32_float32
                    0.05%  577.90ms      3000  192.63us  183.17us  418.56us  adj_patch
                    0.05%  555.15ms    109629  5.0630us  1.5040us  177.44us  cupy_multiply__complex_complex64_complex64
                    0.04%  435.13ms      2254  193.05us  2.3680us  744.70us  cupy_take
                    0.03%  346.67ms    114129  3.0370us  2.5270us  98.240us  cupy_power__float32_float_float32
                    0.03%  297.50ms     10500  28.332us     960ns  220.03us  cupy_conjugate__complex64_complex64
                    0.02%  235.93ms       750  314.57us  310.97us  373.18us  cupy_copy__float32_float32
                    0.01%  141.45ms     86455  1.6360us  1.5030us  11.200us  cupy_less_equal__float32_float32_bool
                    0.01%  111.03ms    107434  1.0330us     320ns  787.58us  [CUDA memcpy DtoH]
                    0.01%  90.171ms      3000  30.057us  24.800us  240.19us  cupy_mean
                    0.01%  89.872ms      4500  19.971us  1.4080us  239.10us  cupy_subtract__complex64_complex64_complex64
                    0.01%  82.370ms      6000  13.728us  1.4080us  30.592us  cupy_negative__complex64_complex64
                    0.00%  51.690ms      4500  11.486us  1.5680us  205.12us  cupy_true_divide__complex64_complex64_complex64
                    0.00%  42.123ms      3010  13.994us     703ns  39.472ms  [CUDA memcpy HtoD]
                    0.00%  36.928ms       750  49.237us  47.136us  224.29us  cupy_scatter_update
                    0.00%  7.7255ms      4500  1.7160us  1.2480us  3.4560us  cupy_add__complex64_complex_complex64
                    0.00%  89.053us        49  1.8170us  1.4400us  11.712us  cupy_absolute__float32_float32
                    0.00%  67.648us        49  1.3800us  1.2160us  1.6000us  cupy_less__float32_float_bool
                    0.00%  2.1120us         1  2.1120us  2.1120us  2.1120us  cupy_concatenate_same_size
      API calls:   90.74%  1.0e+03s    107434  9.4366ms  16.883us  59.806ms  cudaMemcpy
                    7.69%  85.8659s   2119942  40.503us  5.8560us  20.863ms  cuLaunchKernel
                    0.46%  5.11546s   7853266     651ns     334ns  2.0435ms  cudaGetDevice
                    0.38%  4.26260s    250760  16.998us  5.9380us  15.350ms  cudaLaunchKernel
                    0.29%  3.26943s    128380  25.466us  11.499us  18.548ms  cudaMemsetAsync
                    0.19%  2.12784s     96215  22.115us  9.8620us  14.777ms  cudaMemcpyAsync
                    0.07%  810.92ms   1253800     646ns     198ns  15.123ms  cuFuncGetAttribute
                    0.05%  534.54ms        19  28.134ms  17.540us  524.04ms  cudaMalloc
                    0.03%  326.40ms         2  163.20ms  3.7150us  326.40ms  cudaFree
                    0.02%  257.61ms         8  32.201ms  17.307us  250.64ms  cudaHostAlloc
                    0.02%  249.91ms    158296  1.5780us     399ns  1.9381ms  cudaSetDevice
                    0.02%  240.18ms        41  5.8579ms  4.8082ms  10.819ms  cuModuleLoadData
                    0.01%  113.46ms      3009  37.705us     820ns  14.624ms  cudaEventDestroy
                    0.01%  98.000ms    125380     781ns     414ns  353.16us  cudaPeekAtLastError
                    0.00%  46.135ms    125380     367ns     175ns  2.0856ms  cudaGetErrorString
                    0.00%  45.177ms    125380     360ns     190ns  316.96us  cudaGetLastError
                    0.00%  27.950ms      3010  9.2850us  2.8280us  10.986ms  cudaEventRecord
                    0.00%  24.893ms      3010  8.2700us  1.6900us  14.447ms  cudaEventCreateWithFlags
                    0.00%  17.460ms        11  1.5873ms  216.70us  13.418ms  cuModuleUnload
                    0.00%  15.388ms      5945  2.5880us     716ns  327.84us  cudaEventQuery
                    0.00%  800.25us         2  400.12us  371.91us  428.34us  cuDeviceTotalMem
                    0.00%  638.68us       191  3.3430us     299ns  154.84us  cuDeviceGetAttribute
                    0.00%  524.31us         2  262.16us  217.98us  306.34us  cudaGetDeviceProperties
                    0.00%  230.96us         2  115.48us  40.288us  190.67us  cuDeviceGetName
                    0.00%  221.28us         2  110.64us  87.374us  133.91us  cudaMemGetInfo
                    0.00%  81.860us        42  1.9490us  1.0770us  3.1200us  cuModuleGetFunction
                    0.00%  44.774us        32  1.3990us     578ns  5.1140us  cudaFuncSetCacheConfig
                    0.00%  10.804us         7  1.5430us     301ns  5.8330us  cudaGetDeviceCount
                    0.00%  8.6740us         1  8.6740us  8.6740us  8.6740us  cudaDeviceCanAccessPeer
                    0.00%  5.6900us         1  5.6900us  5.6900us  5.6900us  cuDeviceGetPCIBusId
                    0.00%  3.8120us         2  1.9060us     888ns  2.9240us  cudaDeviceGetAttribute
                    0.00%  3.3110us         4     827ns     474ns  1.6730us  cuDeviceGetCount
                    0.00%  2.3860us         1  2.3860us  2.3860us  2.3860us  cudaRuntimeGetVersion
                    0.00%  2.0590us         3     686ns     376ns  1.2850us  cuDeviceGet
                    0.00%  1.7780us         1  1.7780us  1.7780us  1.7780us  cuInit
                    0.00%  1.3750us         1  1.3750us  1.3750us  1.3750us  cuDriverGetVersion
                    0.00%  1.3690us         2     684ns     669ns     700ns  cuDeviceGetUuid
```
Here is the profiling result from `lstsq_grad`.
```
            Type  Time(%)      Time     Calls       Avg       Min       Max  Name
 GPU activities:   19.44%  6.71167s      4503  1.4905ms  16.224us  5.2621ms  cupy_sum
                   13.28%  4.58700s      3750  1.2232ms  1.6320us  1.9018ms  cupy_copy__float32_float32
                   11.13%  3.84159s       750  5.1221ms  5.0548ms  6.3462ms  void gemmSN_NN_kernel<float, int=256, int=4, int=2, int=8, int=2, int=4, cublasGemvTensorStridedBatched<float const >, cublasGemvTensorStridedBatched<float>>(cublasGemmSmallNParams<float const , cublasGemvTensorStridedBatched<float const >, float>)
                    8.35%  2.88492s       750  3.8466ms  3.8102ms  5.1551ms  void gemmSN_NN_kernel<float, int=128, int=2, int=4, int=8, int=2, int=4, cublasGemvTensorStridedBatched<float const >, cublasGemvTensorStridedBatched<float>>(cublasGemmSmallNParams<float const , cublasGemvTensorStridedBatched<float const >, float>)
                    4.07%  1.40633s      2250  625.04us  616.35us  678.24us  cupy_copy__complex64_complex64
                    4.05%  1.39781s      1500  931.87us  903.49us  1.2532ms  cupy_subtract__complex64_complex64_complex64
                    2.82%  973.71ms      2250  432.76us  36.192us  979.42us  cupy_multiply__float32_complex64_complex64
                    2.77%  957.37ms      1501  637.82us  339.81us  821.21us  void vector_fft<unsigned int=256, unsigned int=16, unsigned int=1, padding_t=6, twiddle_t=0, loadstore_modifier_t=2, layout_t=0, unsigned int, float>(kernel_arguments_t<unsigned int>)
                    2.77%  954.99ms      4504  212.03us  1.9840us  368.51us  cupy_sqrt__float32_float32
                    2.73%  942.53ms      1501  627.93us  359.13us  679.65us  cupy_true_divide__complex64_complex_complex64
                    2.41%  830.97ms      4501  184.62us  22.688us  1.0835ms  cupy_multiply__complex64_complex64_complex64
                    2.30%  794.66ms       750  1.0595ms  1.0281ms  1.4850ms  void gemv2T_kernel_val<int, int, float, float, float, int=128, int=16, int=2, int=2, bool=0, cublasGemvParams<cublasGemvTensorStridedBatched<float const >, cublasGemvTensorStridedBatched<float>, float>>(float const , float, float)
                    2.15%  741.28ms      2250  329.46us  1.1840us  1.2048ms  cupy_add__complex64_complex64_complex64
                    2.07%  715.94ms      3001  238.57us  1.3760us  507.10us  cupy_absolute__complex64_float32
                    1.97%  680.87ms      3751  181.52us  11.968us  348.99us  [CUDA memset]
                    1.85%  638.74ms      2251  283.76us  1.4720us  1.2422ms  cupy_multiply__complex64_float32_complex64
                    1.54%  530.34ms       750  707.12us  647.07us  998.11us  void spRadix0256C::kernel1Mem<unsigned int, float, fftDirection_t=1, unsigned int=16, unsigned int=2, L1, ALL, WRITEBACK>(kernel_parameters_t<fft_mem_radix1_t, unsigned int, float>)
                    1.52%  523.55ms       751  697.14us  377.86us  1.0913ms  void spRadix0256C::kernel1Mem<unsigned int, float, fftDirection_t=-1, unsigned int=16, unsigned int=2, L1, ALL, WRITEBACK>(kernel_parameters_t<fft_mem_radix1_t, unsigned int, float>)
                    1.43%  492.10ms      2252  218.52us  21.504us  362.43us  cupy_multiply__float32_float32_float32
                    1.41%  486.64ms      1500  324.43us  1.8240us  1.0398ms  cupy_conjugate__complex64_complex64
                    1.36%  470.98ms       750  627.98us  622.14us  927.04us  cupy_add__complex_complex64_complex64
                    1.36%  470.01ms       750  626.68us  622.24us  666.01us  cupy_multiply__complex_complex64_complex64
                    1.24%  429.41ms      2250  190.85us  183.68us  272.38us  adj_patch
                    1.16%  401.61ms      1503  267.21us  2.3040us  743.65us  cupy_take
                    1.01%  347.26ms       800  434.07us  2.0480us  510.97us  cupy_true_divide__float32_float32_float32
                    1.00%  344.26ms       799  430.86us  1.8880us  505.18us  cupy_subtract__float32_float32_float32
                    0.73%  252.34ms      2251  112.10us  1.1520us  480.00us  cupy_square__float32_float32
                    0.69%  237.89ms      1500  158.60us  1.2800us  341.57us  cupy_add__float32_float_float32
                    0.68%  235.85ms       750  314.47us  311.26us  341.25us  cupy_subtract__float_float32_float32
                    0.31%  106.28ms      2251  47.215us  35.328us  59.584us  fwd_patch
                    0.16%  56.547ms      1500  37.698us  35.840us  43.968us  cupy_true_divide__complex64_complex64_complex64
                    0.12%  42.567ms      1510  28.190us     672ns  39.790ms  [CUDA memcpy HtoD]
                    0.05%  17.478ms       750  23.304us  22.464us  27.168us  cupy_add__complex64_complex_complex64
                    0.03%  10.522ms      1500  7.0140us  1.3430us  154.94us  fill
                    0.01%  3.1559ms      1555  2.0290us     736ns  847.74us  [CUDA memcpy DtoH]
                    0.01%  2.6873ms      1500  1.7910us  1.4080us  2.5920us  cupy_true_divide__complex64_float32_complex64
                    0.01%  2.4929ms       750  3.3230us  2.7520us  293.25us  cupy_power__float32_float_float32
                    0.01%  2.0538ms       750  2.7380us  2.6560us  3.5520us  void getrf_2x2<float, float, int=256, bool=1>(int, float* const *, int, int*, int*, int)
                    0.01%  2.0176ms       750  2.6900us  2.6240us  3.5840us  void getri_2x2<float, int=256, bool=1>(int, float const * const *, int, int const *, float* const *, int, int*, int)
                    0.00%  1.5868ms      1500  1.0570us     736ns  9.1840us  cupy_arange__uint_uint_uint64
                    0.00%  171.81us         1  171.81us  171.81us  171.81us  cupy_add__float_float32_float32
                    0.00%  81.152us        49  1.6560us  1.6000us  1.8560us  cupy_absolute__float32_float32
                    0.00%  75.040us        49  1.5310us  1.3760us  1.7280us  cupy_less__float32_float_bool
                    0.00%  2.1120us         1  2.1120us  2.1120us  2.1120us  cupy_concatenate_same_size
      API calls:   90.75%  30.8058s      1556  19.798ms  24.436us  34.490ms  cudaMemcpy
                    2.35%  796.52ms         9  88.502ms  1.3080us  337.97ms  cudaFree
                    2.27%  770.67ms     57217  13.469us  6.8040us  820.64us  cuLaunchKernel
                    1.70%  577.20ms        30  19.240ms  8.3470us  561.47ms  cudaMalloc
                    0.86%  291.35ms         7  41.622ms  19.596us  284.71ms  cudaHostAlloc
                    0.84%  284.04ms        44  6.4554ms  5.0455ms  18.059ms  cuModuleLoadData
                    0.51%  174.39ms    225883     772ns     357ns  718.79us  cudaGetDevice
                    0.26%  88.870ms      6752  13.162us  6.9110us  75.277us  cudaLaunchKernel
                    0.22%  73.845ms      3751  19.686us  12.953us  81.712us  cudaMemsetAsync
                    0.08%  26.428ms      1509  17.513us  12.180us  87.292us  cudaMemcpyAsync
                    0.06%  18.789ms     45010     417ns     226ns  617.07us  cuFuncGetAttribute
                    0.03%  9.8472ms      4417  2.2290us     694ns  27.628us  cudaEventQuery
                    0.02%  6.0748ms      1509  4.0250us  3.0560us  53.754us  cudaEventRecord
                    0.01%  4.1071ms        12  342.25us  155.63us  742.89us  cuModuleUnload
                    0.01%  4.0866ms      1527  2.6760us     687ns  18.803us  cudaEventCreateWithFlags
                    0.01%  3.5945ms      1526  2.3550us     522ns  19.538us  cudaEventDestroy
                    0.01%  2.1879ms      1570  1.3930us     513ns  13.035us  cudaSetDevice
                    0.01%  2.0090ms      5251     382ns     208ns  328.90us  cudaGetLastError
                    0.01%  1.9273ms         4  481.81us  337.61us  575.11us  cuDeviceTotalMem
                    0.00%  1.1161ms      1501     743ns     490ns  16.181us  cudaPeekAtLastError
                    0.00%  1.0090ms       379  2.6620us     206ns  132.06us  cuDeviceGetAttribute
                    0.00%  529.87us         2  264.94us  209.29us  320.59us  cudaGetDeviceProperties
                    0.00%  514.68us      1501     342ns     212ns  14.914us  cudaGetErrorString
                    0.00%  209.47us         4  52.366us  38.402us  69.480us  cuDeviceGetName
                    0.00%  97.175us         1  97.175us  97.175us  97.175us  cudaMemGetInfo
                    0.00%  93.124us        45  2.0690us  1.3050us  2.6760us  cuModuleGetFunction
                    0.00%  46.766us        32  1.4610us     578ns  5.0220us  cudaFuncSetCacheConfig
                    0.00%  36.963us         4  9.2400us  1.7570us  16.890us  cudaDeviceSynchronize
                    0.00%  11.630us        13     894ns     448ns  3.2340us  cudaDeviceGetAttribute
                    0.00%  6.1120us         1  6.1120us  6.1120us  6.1120us  cuDeviceGetPCIBusId
                    0.00%  5.0770us         3  1.6920us  1.4430us  1.9510us  cuInit
                    0.00%  5.0130us         5  1.0020us     286ns  1.8230us  cudaGetDeviceCount
                    0.00%  4.8240us         6     804ns     381ns  2.1630us  cuDeviceGetCount
                    0.00%  3.1020us         5     620ns     311ns  1.3310us  cuDeviceGet
                    0.00%  2.9710us         3     990ns     725ns  1.3300us  cuDriverGetVersion
                    0.00%  2.5120us         1  2.5120us  2.5120us  2.5120us  cudaRuntimeGetVersion
                    0.00%  2.3260us         4     581ns     374ns     766ns  cuDeviceGetUuid
```

**Review by xiaodong-yu — APPROVED** [review]

**Pull request merged**